binding.go

```diff
@@ -201,13 +204,27 @@ func Json(jsonStruct interface{}, ifacePtr ...interface{}) macaron.Handler {
 		var errors Errors
 		ensureNotPointer(jsonStruct)
 		jsonStruct := reflect.New(reflect.TypeOf(jsonStruct))
-		if ctx.Req.Request.Body != nil {
-			defer ctx.Req.Request.Body.Close()
-			err := json.NewDecoder(ctx.Req.Request.Body).Decode(jsonStruct.Interface())
-			if err != nil && err != io.EOF {
-				errors.Add([]string{}, ERR_DESERIALIZATION, err.Error())
+		var err error
+		if ctx.Req.Method == "POST" || ctx.Req.Method == "PUT" || ctx.Req.Method == "PATCH" {
+			if ctx.Req.Request.Body != nil {
+				v := jsonStruct.Interface()
+				if pb, ok := v.(proto.Message); ok {
+					err = jsonpb.Unmarshal(ctx.Req.Request.Body, pb)
+				} else {
+					err = json.NewDecoder(ctx.Req.Request.Body).Decode(v)
+				}
 			}
 		}
+		if (err == nil || err == io.EOF) && ctx.Req.URL != nil {
```

Comment:
OK, how about this way.

1. Use `else-if ctx.Req.Method == "GET"` to indicate this is another branch of conditions, do not use `err == nil`, feels too hacky.
2. Add comment for this block of code explain the reason just as you did in your comment.

Reply:
I have updated the pr to match ParseForm() more closely and added comments. PTAL.